fdbserver/workloads/RemoveServersSafely.actor.cpp

```diff
@@ -401,11 +402,70 @@ struct RemoveServersSafelyWorkload : TestWorkload {
 
 		state std::vector<ISimulator::ProcessInfo*>	killProcArray;
 		state std::vector<AddressExclusion>	toKillArray;
-
+		state std::vector<AddressExclusion>	toKillMarkFailedArray;
+		state AddressExclusion coordExcl;
+		// Exclude a coordinator under buggify, but only if fault tolerance is > 0 and kill set is non-empty already
+		if (BUGGIFY && toKill.size()) {
+			std::vector<NetworkAddress> coordinators = wait(getCoordinators(cx));
+			if (coordinators.size() > 2) {
+				auto randomCoordinator = deterministicRandom()->randomChoice(coordinators);
+				coordExcl = AddressExclusion(randomCoordinator.ip, randomCoordinator.port);
+			}
+		}
 		std::copy(toKill.begin(), toKill.end(), std::back_inserter(toKillArray));
+		if (markExcludeAsFailed) {
+			state int retries = 0;
+			loop {
+				state bool safe = false;
+				state std::set<AddressExclusion> failSet =
+				    random_subset(toKillArray, deterministicRandom()->randomInt(0, toKillArray.size() + 1));
+				if (coordExcl.isValid()) {
+					failSet.insert(coordExcl);
+				}
+				toKillMarkFailedArray.resize(failSet.size());
+				std::copy(failSet.begin(), failSet.end(), toKillMarkFailedArray.begin());
+				TraceEvent("RemoveAndKill", functionId)
+				    .detail("Step", "SafetyCheck")
+				    .detail("Exclusions", describe(toKillMarkFailedArray));
+				choose {
+					when(bool _safe = wait(checkSafeExclusions(cx, toKillMarkFailedArray))) {
+						safe = _safe;
+					}
+					when(wait(delay(5.0))) {
+						TraceEvent("RemoveAndKill", functionId)
+							.detail("Step", "SafetyCheckTimedOut")
+							.detail("Exclusions", describe(toKillMarkFailedArray));
+					}
+				}
+				if (retries == self->maxSafetyCheckRetries) {
+					// Do not mark as failed if limit is reached
+					TraceEvent("RemoveAndKill", functionId)
+						.detail("Step", "SafetyCheckLimitReached")
+						.detail("Retries", retries);
+					markExcludeAsFailed = false;
+					safe = true;
+				}
+				if (safe) break;
+				retries++;
+			}
+		}
+		// Swap out coordinator with server in kill set, but only if already marking as failed and safety check passes
```

Comment:
Since the `toKill` servers plus the `corrdExcl` have passed the safety check, why do we still want to remove a server from the kill set?

What will happen if we do not do `toKill.erase(removeServer);` or `			toKillArray.erase(std::remove(toKillArray.begin(), toKillArray.end(), *removeServer), toKillArray.end());`?

Reply:
Same as above, this is a mechanism to make sure the number of kills stays consistent. The safety check for failed servers is a subset of the number of servers to be excluded, so passing the safety check does not guarantee that the exclusion of N + 1 servers keeps the state of the db stable. 

For a more concrete example, say the db can tolerate 4 processes dying, so the workload creates a kill set of size 4. The safety check takes a subset of say 2 of these servers, and then adds our extra coordinator and our check passes. If we now add the coordinator without replacement, I think it is possible (but not guaranteed) that the exclusion of those 4 servers + 1 extra coordinator may leave the db in a bad state.

Reply:
Doesn't the set of processes to kill that passes the safetyCheck already include the coordinator?
See https://github.com/apple/foundationdb/blob/f4237ebffffad14f370fc1c99ac241bbe3c086d9/fdbserver/workloads/RemoveServersSafely.actor.cpp#L423, the `coordExcl ` is added to `failSet ` and the `failSet` is passed into `checkSafeExclusions `.

Did I miss something?

Reply:
Yes, the coordinator is checked, but the thing I am referring to is the number of servers being killed. It has nothing to do with the process classes. The main points in the example given are here:

- DB can only tolerate 4 kills, picks a set of 4 servers to kill
- Safety check performs the check on a set of 3, which is a random 2 out of the 4 chosen, plus the coordinator
- Safety check passes for our set of 3 kills
- Now, seeing the check has passed, we add the coordinator to the kill list, creating 5 kills total. This could break the DB unless we shrink our kill list back down to 4.

Reply:
I see the reason now. 
Can you change the comment `		// Swap out coordinator with server in kill set, but only if already marking as failed and safety check passes
` to `		// Swap coordinator with one server in the kill set to ensure the number of processes to kill does not increase. This is needed only if a new coordinator is added to the toKill set in this function and safety check passes`